Generate: AssertionError: null: com.graphicsfuzz.common.typing.Typer.visitArrayIndexExpr(Typer.java:503)

To reproduce:

`graphicsfuzz-tool com.graphicsfuzz.generator.tool.Generate crash_generate/shader.json donors 100 variant.json --generate-uniform-bindings --max-uniforms 10 --seed=-173811911`

[AssertionError.zip](https://github.com/google/graphicsfuzz/files/3294123/AssertionError.zip)

<details>
 <summary>Stacktrace:</summary>

```
$ graphicsfuzz-tool com.graphicsfuzz.generator.tool.Generate crash_generate/shader.json donors 100 variant.json --generate-uniform-bindings --max-uniforms 10 --seed=-173811911
/data/git/graphicsfuzz/graphicsfuzz/target/graphicsfuzz/jar/tool-1.0.jar
['java', '-ea', '-cp', '/data/git/graphicsfuzz/graphicsfuzz/target/graphicsfuzz/jar/tool-1.0.jar', 'com.graphicsfuzz.generator.tool.Generate', 'crash_generate/shader.json', 'donors', '100', 'variant.json', '--generate-uniform-bindings', '--max-uniforms', '10', '--seed=-173811911']
13:59:01.222 INFO  ExecHelper - /data/git/graphicsfuzz/graphicsfuzz/target/graphicsfuzz/bin/Linux/glslangValidator -E --stdin -S frag
13:59:01.281 INFO  ExecHelper - Result: 0
13:59:01.648 ERROR Generate - 
java.lang.AssertionError: null
	at com.graphicsfuzz.common.typing.Typer.visitArrayIndexExpr(Typer.java:503) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.ArrayIndexExpr.accept(ArrayIndexExpr.java:53) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitTypeConstructorExpr(StandardVisitor.java:235) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitTypeConstructorExpr(Typer.java:182) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.TypeConstructorExpr.accept(TypeConstructorExpr.java:86) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitFunctionCallExpr(StandardVisitor.java:222) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitFunctionCallExpr(Typer.java:112) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.FunctionCallExpr.accept(FunctionCallExpr.java:73) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitFunctionCallExpr(StandardVisitor.java:222) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitFunctionCallExpr(Typer.java:112) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.FunctionCallExpr.accept(FunctionCallExpr.java:73) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitBinaryExpr(StandardVisitor.java:178) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitBinaryExpr(Typer.java:302) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.BinaryExpr.accept(BinaryExpr.java:65) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitFunctionCallExpr(StandardVisitor.java:222) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitFunctionCallExpr(Typer.java:112) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.FunctionCallExpr.accept(FunctionCallExpr.java:73) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitParenExpr(StandardVisitor.java:183) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitParenExpr(Typer.java:83) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.ParenExpr.accept(ParenExpr.java:36) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitArrayIndexExpr(StandardVisitor.java:325) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitArrayIndexExpr(Typer.java:491) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.ArrayIndexExpr.accept(ArrayIndexExpr.java:53) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitFunctionCallExpr(StandardVisitor.java:222) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitFunctionCallExpr(Typer.java:112) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.FunctionCallExpr.accept(FunctionCallExpr.java:73) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitFunctionCallExpr(StandardVisitor.java:222) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitFunctionCallExpr(Typer.java:112) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.FunctionCallExpr.accept(FunctionCallExpr.java:73) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitParenExpr(StandardVisitor.java:183) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitParenExpr(Typer.java:83) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.ParenExpr.accept(ParenExpr.java:36) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitBinaryExpr(StandardVisitor.java:177) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitBinaryExpr(Typer.java:302) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.BinaryExpr.accept(BinaryExpr.java:65) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitFunctionCallExpr(StandardVisitor.java:222) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitFunctionCallExpr(Typer.java:112) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.FunctionCallExpr.accept(FunctionCallExpr.java:73) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitFunctionCallExpr(StandardVisitor.java:222) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitFunctionCallExpr(Typer.java:112) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.FunctionCallExpr.accept(FunctionCallExpr.java:73) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitFunctionCallExpr(StandardVisitor.java:222) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitFunctionCallExpr(Typer.java:112) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.FunctionCallExpr.accept(FunctionCallExpr.java:73) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitTypeConstructorExpr(StandardVisitor.java:235) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitTypeConstructorExpr(Typer.java:182) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.TypeConstructorExpr.accept(TypeConstructorExpr.java:86) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitFunctionCallExpr(StandardVisitor.java:222) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitFunctionCallExpr(Typer.java:112) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.FunctionCallExpr.accept(FunctionCallExpr.java:73) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitFunctionCallExpr(StandardVisitor.java:222) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitFunctionCallExpr(Typer.java:112) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.FunctionCallExpr.accept(FunctionCallExpr.java:73) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitParenExpr(StandardVisitor.java:183) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitParenExpr(Typer.java:83) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.ParenExpr.accept(ParenExpr.java:36) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitBinaryExpr(StandardVisitor.java:177) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitBinaryExpr(Typer.java:302) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.BinaryExpr.accept(BinaryExpr.java:65) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitFunctionCallExpr(StandardVisitor.java:222) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitFunctionCallExpr(Typer.java:112) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.FunctionCallExpr.accept(FunctionCallExpr.java:73) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitArrayIndexExpr(StandardVisitor.java:325) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitArrayIndexExpr(Typer.java:491) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.ArrayIndexExpr.accept(ArrayIndexExpr.java:53) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitTypeConstructorExpr(StandardVisitor.java:235) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitTypeConstructorExpr(Typer.java:182) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.TypeConstructorExpr.accept(TypeConstructorExpr.java:86) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitFunctionCallExpr(StandardVisitor.java:222) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitFunctionCallExpr(Typer.java:112) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.FunctionCallExpr.accept(FunctionCallExpr.java:73) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitFunctionCallExpr(StandardVisitor.java:222) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitFunctionCallExpr(Typer.java:112) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.FunctionCallExpr.accept(FunctionCallExpr.java:73) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitTernaryExpr(StandardVisitor.java:296) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitTernaryExpr(Typer.java:288) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.TernaryExpr.accept(TernaryExpr.java:54) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitParenExpr(StandardVisitor.java:183) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitParenExpr(Typer.java:83) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.ParenExpr.accept(ParenExpr.java:36) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitFunctionCallExpr(StandardVisitor.java:222) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitFunctionCallExpr(Typer.java:112) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.FunctionCallExpr.accept(FunctionCallExpr.java:73) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitMemberLookupExpr(StandardVisitor.java:197) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitMemberLookupExpr(Typer.java:394) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.MemberLookupExpr.accept(MemberLookupExpr.java:53) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitBinaryExpr(StandardVisitor.java:177) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitBinaryExpr(Typer.java:302) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.BinaryExpr.accept(BinaryExpr.java:65) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitBinaryExpr(StandardVisitor.java:178) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitBinaryExpr(Typer.java:302) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.expr.BinaryExpr.accept(BinaryExpr.java:65) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitIfStmt(StandardVisitor.java:138) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.stmt.IfStmt.accept(IfStmt.java:65) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:419) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitBlockStmt(StandardVisitor.java:132) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.ScopeTreeBuilder.visitBlockStmt(ScopeTreeBuilder.java:67) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.stmt.BlockStmt.accept(BlockStmt.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.ScopeTreeBuilder.visitFunctionDefinition(ScopeTreeBuilder.java:140) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.visitFunctionDefinition(Typer.java:102) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.decl.FunctionDefinition.accept(FunctionDefinition.java:42) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitChildFromParent(StandardVisitor.java:415) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visitTranslationUnit(StandardVisitor.java:113) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.TranslationUnit.accept(TranslationUnit.java:123) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.ast.visitors.StandardVisitor.visit(StandardVisitor.java:98) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.common.typing.Typer.<init>(Typer.java:74) ~[ast-1.0.jar:?]
	at com.graphicsfuzz.generator.tool.Generate.canTypeCheckWithoutFailure(Generate.java:456) ~[generator-1.0.jar:?]
	at com.graphicsfuzz.generator.tool.Generate.applyTransformationsMultiPass(Generate.java:438) ~[generator-1.0.jar:?]
	at com.graphicsfuzz.generator.tool.Generate.transformShader(Generate.java:291) ~[generator-1.0.jar:?]
	at com.graphicsfuzz.generator.tool.Generate.generateVariant(Generate.java:184) ~[generator-1.0.jar:?]
	at com.graphicsfuzz.generator.tool.Generate.generateVariant(Generate.java:233) ~[generator-1.0.jar:?]
	at com.graphicsfuzz.generator.tool.Generate.mainHelper(Generate.java:333) ~[generator-1.0.jar:?]
	at com.graphicsfuzz.generator.tool.Generate.main(Generate.java:313) [generator-1.0.jar:?]
```

</details>

--- a/generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java
+++ b/generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java
@@ -105,7 +105,10 @@ public final class OpaqueExpressionGenerator {
     );
   }
 
-  private List<OpaqueZeroOneFactory> waysToMakeZero() {
+  /**
+   * This method has non-private visibility for purposes of testing only.
+   */
+  List<OpaqueZeroOneFactory> waysToMakeZero() {
     List<OpaqueZeroOneFactory> opaqueZeroFactories = new ArrayList<>();
     opaqueZeroFactories.addAll(waysToMakeZeroOrOne());
     opaqueZeroFactories.add(this::opaqueZeroSin);
@@ -115,7 +118,10 @@ public final class OpaqueExpressionGenerator {
     return opaqueZeroFactories;
   }
 
-  private List<OpaqueZeroOneFactory> waysToMakeOne() {
+  /**
+   * This method has non-private visibility for purposes of testing only.
+   */
+  List<OpaqueZeroOneFactory> waysToMakeOne() {
     List<OpaqueZeroOneFactory> opaqueOneFactories = new ArrayList<>();
     opaqueOneFactories.addAll(waysToMakeZeroOrOne());
     opaqueOneFactories.add(this::opaqueOneExponential);
@@ -402,10 +408,17 @@ public final class OpaqueExpressionGenerator {
                                                 Fuzzer fuzzer, boolean isZero) {
     // represent 0 as the length of zero vector, e.g. length(opaqueZero).
     assert isZero;
-    if (!BasicType.allGenTypes().contains(type)) {
+    if (type != BasicType.FLOAT) {
+      // 'length' has return type 'float', so we can only create a scalar floating-point zero.
       return Optional.empty();
     }
-    return Optional.of(new FunctionCallExpr("length", makeOpaqueZero(type, constContext, depth,
+
+    // We can choose any vector size we wish for the vector whose length will be computed.
+    final BasicType vectorType =
+        BasicType.allGenTypes().get(generator.nextInt(BasicType.allGenTypes().size()));
+
+    return Optional.of(new FunctionCallExpr("length", makeOpaqueZero(vectorType, constContext,
+        depth,
         fuzzer)));
   }
 
@@ -434,13 +447,21 @@ public final class OpaqueExpressionGenerator {
   private Optional<Expr> opaqueOneNormalizedVectorLength(BasicType type, boolean constContext,
                                                          final int depth,
                                                          Fuzzer fuzzer, boolean isZero) {
-    // represent 1 as the length of normalized vector
+    // represent 1 as the length of a normalized vector
     assert !isZero;
-    if (!BasicType.allGenTypes().contains(type)) {
+    if (type != BasicType.FLOAT) {
+      // 'length' has return type 'float', so we can only create a scalar floating-point zero.
       return Optional.empty();
     }
-    Expr normalizedExpr = new FunctionCallExpr("normalize", makeOpaqueZeroOrOne(isZero,
-        type, constContext, depth, fuzzer));
+
+    // We can choose any vector size we wish for the vector whose length will be computed.
+    final BasicType vectorType =
+        BasicType.allGenTypes().get(generator.nextInt(BasicType.allGenTypes().size()));
+
+    // We create a vector of ones and normalize it.  Note that we could be more general and create
+    // any non-zero vector and normalize it.
+    Expr normalizedExpr = new FunctionCallExpr("normalize", makeOpaqueZeroOrOne(false,
+        vectorType, constContext, depth, fuzzer));
     return Optional.of(new FunctionCallExpr("length", normalizedExpr));
   }
 

--- a/generator/src/test/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGeneratorTest.java
+++ b/generator/src/test/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGeneratorTest.java
@@ -20,6 +20,7 @@ import com.graphicsfuzz.common.ast.TranslationUnit;
 import com.graphicsfuzz.common.ast.decl.FunctionDefinition;
 import com.graphicsfuzz.common.ast.decl.FunctionPrototype;
 import com.graphicsfuzz.common.ast.decl.PrecisionDeclaration;
+import com.graphicsfuzz.common.ast.decl.ScalarInitializer;
 import com.graphicsfuzz.common.ast.decl.VariableDeclInfo;
 import com.graphicsfuzz.common.ast.decl.VariablesDeclaration;
 import com.graphicsfuzz.common.ast.expr.BinOp;
@@ -31,12 +32,16 @@ import com.graphicsfuzz.common.ast.stmt.DeclarationStmt;
 import com.graphicsfuzz.common.ast.stmt.ExprStmt;
 import com.graphicsfuzz.common.ast.stmt.Stmt;
 import com.graphicsfuzz.common.ast.type.BasicType;
+import com.graphicsfuzz.common.ast.type.QualifiedType;
+import com.graphicsfuzz.common.ast.type.Type;
+import com.graphicsfuzz.common.ast.type.TypeQualifier;
 import com.graphicsfuzz.common.ast.type.VoidType;
 import com.graphicsfuzz.common.glslversion.ShadingLanguageVersion;
 import com.graphicsfuzz.common.transformreduce.GlslShaderJob;
 import com.graphicsfuzz.common.typing.Scope;
 import com.graphicsfuzz.common.typing.SupportedTypes;
 import com.graphicsfuzz.common.util.IRandom;
+import com.graphicsfuzz.common.util.IdGenerator;
 import com.graphicsfuzz.common.util.PipelineInfo;
 import com.graphicsfuzz.common.util.RandomWrapper;
 import com.graphicsfuzz.common.util.ShaderJobFileOperations;
@@ -46,6 +51,7 @@ import com.graphicsfuzz.generator.util.GenerationParams;
 import java.io.File;
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Collections;
 import java.util.List;
 import java.util.Optional;
 import org.junit.Rule;
@@ -61,25 +67,34 @@ public class OpaqueExpressionGeneratorTest {
 
   @Test
   public void testGeneratesValidExpressions() throws Exception {
+    // We make a random number generator here and pass it into functions below in order to maximise
+    // diversity - even though there are differences between shading languages, there is still a
+    // good chance that until such a difference is hit, generating expressions for different
+    // shading languages but from the same seed will lead to identical results.
+    final IRandom generator = new RandomWrapper(0);
+
+    // These are the shading language versions we support best, so test them thoroughly.
     for (ShadingLanguageVersion shadingLanguageVersion : Arrays.asList(
+        ShadingLanguageVersion.ESSL_100,
         ShadingLanguageVersion.ESSL_300,
-        ShadingLanguageVersion.ESSL_100)) {
+        ShadingLanguageVersion.ESSL_310,
+        ShadingLanguageVersion.ESSL_320)) {
       for (BasicType t : BasicType.allBasicTypes()) {
         if (!SupportedTypes.supported(t, shadingLanguageVersion)) {
           continue;
         }
-        final TranslationUnit tu = new TranslationUnit(Optional.of(ShadingLanguageVersion.ESSL_310),
+        final TranslationUnit tu = new TranslationUnit(Optional.of(shadingLanguageVersion),
             Arrays.asList(
                 new PrecisionDeclaration("precision mediump float;"),
                 new FunctionDefinition(
-                    new FunctionPrototype("main", VoidType.VOID, new ArrayList<>()),
+                    new FunctionPrototype("main", VoidType.VOID, Collections.emptyList()),
                     new BlockStmt(makeMutatedExpressionAssignments(t,
-                        shadingLanguageVersion, 1000),
+                        shadingLanguageVersion, 1000, generator),
                         false))));
         Generate.addInjectionSwitchIfNotPresent(tu);
         final File shaderJobFile = temporaryFolder.newFile("ex.json");
         final ShaderJobFileOperations fileOps = new ShaderJobFileOperations();
-        fileOps.writeShaderJobFile(new GlslShaderJob(Optional.empty(), new PipelineInfo("{}"),
+        fileOps.writeShaderJobFile(new GlslShaderJob(Optional.empty(), new PipelineInfo(),
             tu), shaderJobFile);
         assertTrue(fileOps.areShadersValid(shaderJobFile, false));
         fileOps.deleteShaderJobFile(shaderJobFile);
@@ -89,7 +104,8 @@ public class OpaqueExpressionGeneratorTest {
 
   private List<Stmt> makeMutatedExpressionAssignments(BasicType basicType,
                                                       ShadingLanguageVersion shadingLanguageVersion,
-                                                      int numberOfAssignments) {
+                                                      int numberOfAssignments,
+                                                      IRandom generator) {
     // We declare only one variable and generate a number of assignments to this variable,
     // instead of making multiple declarations and assignments.
     // numberOfAssignments should be large enough to get high coverage, e.g. 1000.
@@ -99,7 +115,6 @@ public class OpaqueExpressionGeneratorTest {
         new VariableDeclInfo("x", null,
             null))));
     for (int i = 0; i < numberOfAssignments; i++) {
-      final IRandom generator = new RandomWrapper(i);
       final OpaqueExpressionGenerator opaqueExpressionGenerator =
           new OpaqueExpressionGenerator(generator,
               generationParams, shadingLanguageVersion);
@@ -114,4 +129,80 @@ public class OpaqueExpressionGeneratorTest {
     }
     return newStmts;
   }
+
+  @Test
+  public void testWaysToMakeZeroAndOne() throws Exception {
+
+    final IRandom generator = new RandomWrapper(0);
+    final GenerationParams generationParams = GenerationParams.large(ShaderKind.FRAGMENT, true);
+
+    // These are the shading language versions we support best, so test them thoroughly.
+    for (ShadingLanguageVersion shadingLanguageVersion : Arrays.asList(
+        ShadingLanguageVersion.ESSL_100,
+        ShadingLanguageVersion.ESSL_300,
+        ShadingLanguageVersion.ESSL_310,
+        ShadingLanguageVersion.ESSL_320)) {
+      final IdGenerator idGenerator = new IdGenerator();
+      final List<Stmt> stmts = new ArrayList<>();
+      final OpaqueExpressionGenerator opaqueExpressionGenerator =
+          new OpaqueExpressionGenerator(generator,
+              generationParams, shadingLanguageVersion);
+      for (BasicType basicType : BasicType.allNumericTypes()) {
+        if (!SupportedTypes.supported(basicType, shadingLanguageVersion)) {
+          continue;
+        }
+        for (boolean constContext : Arrays.asList(true, false)) {
+          stmts.addAll(makeStatementsFromFactories(generator, generationParams,
+              shadingLanguageVersion,
+              idGenerator, basicType, constContext, opaqueExpressionGenerator.waysToMakeZero(), true));
+          // We do not allow making one for non-square matrices.
+          if (!BasicType.allNonSquareMatrixTypes().contains(basicType)) {
+            stmts.addAll(makeStatementsFromFactories(generator, generationParams,
+                shadingLanguageVersion,
+                idGenerator, basicType, constContext, opaqueExpressionGenerator.waysToMakeOne(),
+                false));
+          }
+        }
+      }
+      final TranslationUnit tu = new TranslationUnit(Optional.of(shadingLanguageVersion),
+          Arrays.asList(
+              new PrecisionDeclaration("precision mediump float;"),
+              new FunctionDefinition(
+                  new FunctionPrototype("main", VoidType.VOID, Collections.emptyList()),
+                  new BlockStmt(stmts,
+                      false))));
+      Generate.addInjectionSwitchIfNotPresent(tu);
+      final File shaderJobFile = temporaryFolder.newFile("ex.json");
+      final ShaderJobFileOperations fileOps = new ShaderJobFileOperations();
+      fileOps.writeShaderJobFile(new GlslShaderJob(Optional.empty(), new PipelineInfo(),
+          tu), shaderJobFile);
+      assertTrue(fileOps.areShadersValid(shaderJobFile, false));
+      fileOps.deleteShaderJobFile(shaderJobFile);
+    }
+  }
+
+  private List<Stmt> makeStatementsFromFactories(IRandom generator, GenerationParams generationParams, ShadingLanguageVersion shadingLanguageVersion, IdGenerator idGenerator, BasicType typeToGenerate, boolean constContext, List<OpaqueZeroOneFactory> factories, boolean makingZero) {
+    List<Stmt> result = new ArrayList<>();
+    for (OpaqueZeroOneFactory factory : factories) {
+      final Optional<Expr> expr = factory.tryMakeOpaque(typeToGenerate, constContext, 0,
+          new Fuzzer(new FuzzingContext(new Scope(null)), shadingLanguageVersion,
+          generator, generationParams), makingZero);
+      if (expr.isPresent()) {
+        final Type baseType = constContext ? new QualifiedType(typeToGenerate,
+            Collections.singletonList(TypeQualifier.CONST)) : typeToGenerate;
+        result.add(new DeclarationStmt(
+            new VariablesDeclaration(
+                baseType,
+                new VariableDeclInfo(
+                    "v" + idGenerator.freshId(),
+                    null,
+                    new ScalarInitializer(expr.get())
+                )
+            )
+        ));
+      }
+    }
+    return result;
+  }
+
 }

